### PR TITLE
JSUI-2735 Added `DynamicHierarchicalFacetSearch`

### PR DESCRIPTION
**Pull request by btaillon**
This PR adds the `DynamicHierarchicalFacetSearch` component.

# Previews
# Search bar
image

# Search results
image

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/controllers/HierarchicalFacetSearchController.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import { IDynamicHierarchicalFacet, IDynamicHierarchicalFacetValue } from '../ui/DynamicHierarchicalFacet/IDynamicHierarchicalFacet';
+import { FacetSearchType, IFacetSearchRequest } from '../rest/Facet/FacetSearchRequest';
+import { IFacetSearchResponse } from '../rest/Facet/FacetSearchResponse';
+import { flatten } from 'underscore';
+
+type Path = string[];
+
+export class HierarchicalFacetSearchController {
```
Comment (btaillon) [inline src/controllers/HierarchicalFacetSearchController.ts:8]: This class is for the most part a copy of [`FacetSearchController`](https://github.com/coveo/search-ui/blob/master/src/controllers/FacetSearchController.ts). The differences are mostly in the `search` method.

File: src/controllers/HierarchicalFacetSearchController.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  private flattenPaths(value: Path[][]): Path[] {
+    return flatten(value, true);
+  }
+
+  public search(terms?: string): Promise<IFacetSearchResponse> {
+    const request: IFacetSearchRequest = {
+      field: this.facet.fieldName,
+      type: FacetSearchType.hierarchical,
+      numberOfValues: this.facet.options.numberOfValues,
+      ignorePaths: this.ignoredPaths,
```
Comment (btaillon) [inline src/controllers/HierarchicalFacetSearchController.ts:28]: `ignorePaths` [doesn't work yet](https://coveord.atlassian.net/browse/SEARCHAPI-4666). Adding this line here would remove the selected path and every children that happens to be already displayed under the facet's suggestions.

For instance, in the following picture:
image

`ignorePaths` would be:
```ts
[
  ['scala'],
  ['scala', 'syntax'],
  ['scala', 'version']
];
```

File: src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts [hunk 3]
```diff
@@ -364,10 +375,25 @@ export class DynamicHierarchicalFacet extends Component implements IDynamicHiera
 
   private updateAppearance() {
     this.header.toggleCollapse(this.isCollapsed);
+    this.toggleSearchDisplay();
     $$(this.element).toggleClass('coveo-dynamic-hierarchical-facet-collapsed', this.isCollapsed);
     $$(this.element).toggleClass('coveo-hidden', !this.values.allFacetValues.length);
   }
 
+  private toggleSearchDisplay() {
```
Comment (btaillon) [inline src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts:383]: This method is copied from [`DynamicFacet`](https://github.com/coveo/search-ui/blob/master/src/ui/DynamicFacet/DynamicFacet.ts#L770)'s.

File: src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts [hunk 4]
```diff
@@ -612,7 +638,13 @@ export class DynamicHierarchicalFacet extends Component implements IDynamicHiera
     };
   }
 
-  private buildFacetHeader() {
```
Comment (btaillon) [inline src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts:615]: This method was shortened in favor of adopting a structure more similar to [`DynamicFacet`](https://github.com/coveo/search-ui/blob/master/src/ui/DynamicFacet/DynamicFacet.ts#L724)'s.

File: src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearch.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import { HierarchicalFacetSearchController } from '../../controllers/HierarchicalFacetSearchController';
+import { Combobox } from '../Combobox/Combobox';
+import { l } from '../../strings/Strings';
+import { IFacetSearchResponse } from '../../rest/Facet/FacetSearchResponse';
+import { IComboboxValue } from '../Combobox/ComboboxValues';
+import { DynamicHierarchicalFacetSearchValue } from './DynamicHierarchicalFacetSearchValue';
+import { DynamicHierarchicalFacetSearchValueRenderer } from './DynamicHierarchicalFacetSearchValueRenderer';
+import 'styling/DynamicHierarchicalFacetSearch/_DynamicHierarchicalFacetSearch';
+
+export class DynamicHierarchicalFacetSearch {
```
Comment (btaillon) [inline src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearch.ts:11]: This class is heavily inspired by [`DynamicFacetSearch`](https://github.com/coveo/search-ui/blob/master/src/ui/DynamicFacetSearch/DynamicFacetSearch.ts).

File: src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValue.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import {
+  IDynamicHierarchicalFacetSearchValueProperties,
+  IDynamicHierarchicalFacet
+} from '../DynamicHierarchicalFacet/IDynamicHierarchicalFacet';
+import { DynamicHierarchicalFacetSearchValueRenderer } from './DynamicHierarchicalFacetSearchValueRenderer';
+import { analyticsActionCauseList } from '../Analytics/AnalyticsActionListMeta';
+
+export class DynamicHierarchicalFacetSearchValue implements IDynamicHierarchicalFacetSearchValueProperties {
```
Comment (btaillon) [inline src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValue.ts:8]: This class is inspired by [`DynamicFacetValue`](https://github.com/coveo/search-ui/blob/master/src/ui/DynamicFacet/DynamicFacetValues/DynamicFacetValue.ts).

File: src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+      l(
+        'SelectValueWithResultCount',
+        this.facetValue.displayValue,
+        l('ResultCount', this.facetValue.numberOfResults, this.facetValue.numberOfResults)
+      ),
+      [...start, ...(end || [])].reverse().join(', ')
+    );
+  }
+
+  public render() {
```
Comment (btaillon) [inline src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts:57]: Overall, a render should look like:
```html
<div class="coveo-dynamic-hierarchical-facet-search-value" aria-label="Select John Doe with 1337 results under People, Canada, Milky Way">
  <header class="coveo-dynamic-hierarchical-facet-search-value-header" aria-hidden="true">
    <span class="coveo-dynamic-hierarchical-facet-search-value-label">John Doe</span>
    <span class="coveo-dynamic-hierarchical-facet-search-value-results-count">(1337)</span>
  </header>
  <ul class="coveo-dynamic-hierarchical-facet-search-value-path" aria-hidden="true">
    <li class="coveo-dynamic-hierarchical-facet-search-value-path-part">Milky Way</li>
    <li class="coveo-dynamic-hierarchical-facet-search-value-path-separator">/</li>
    <li class="coveo-dynamic-hierarchical-facet-search-value-path-ellipsis">...</li>
    <li class="coveo-dynamic-hierarchical-facet-search-value-path-separator">/</li>
    <li class="coveo-dynamic-hierarchical-facet-search-value-path-part">Canada</li>
    <li class="coveo-dynamic-hierarchical-facet-search-value-path-separator">/</li>
    <li class="coveo-dynamic-hierarchical-facet-search-value-path-part">People</li>
  </ul>
</div>
```

File: src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  private get label() {
+    const { start, end } = this.pathToRender;
+    return l(
+      'HierarchicalFacetValueIndentedUnder',
+      l(
+        'SelectValueWithResultCount',
+        this.facetValue.displayValue,
+        l('ResultCount', this.facetValue.numberOfResults, this.facetValue.numberOfResults)
+      ),
+      [...start, ...(end || [])].reverse().join(', ')
```
Comment (btaillon) [inline src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts:53]: The label spells out categories in the reverse order (small to large). Would it be better the other way around?

File: sass/DynamicHierarchicalFacetSearch/_DynamicHierarchicalFacetSearch.scss [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    .coveo-dynamic-hierarchical-facet-search-value-path-part {
+      flex: 0 15 auto;
+      min-width: 10px;
+      @supports (max-width: max-content) {
+        flex: 1 0 0;
+        max-width: max-content;
+      }
+      white-space: nowrap;
+      overflow: hidden;
+      text-overflow: ellipsis;
```
Comment (btaillon) [inline sass/DynamicHierarchicalFacetSearch/_DynamicHierarchicalFacetSearch.scss:36]: Although only the first and last two path parts are displayed, I added this CSS in case they would be too long. This weird bit of CSS adds an ellipsis to longer values first.

For instance:
`I believe this is a longer phrase/hello` could be shortened to `I believe thi.../hello` or `I be.../hell...`.

File: unitTests/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacetTest.ts [hunk 10]
```diff
@@ -290,6 +298,51 @@ export function DynamicHierarchicalFacetTest() {
...
+    });
+
+    it(`when "enableFacetSearch" option is "undefined" and "moreValuesAvailable" is "true"
+    it should hide the search`, () => {
+      initializeComponent();
+      test.cmp.moreValuesAvailable = false;
+      test.cmp.ensureDom();
+
+      expect(searchFeatureDisplayed()).toBe(false);
+    });
```
Comment (btaillon) [inline unitTests/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacetTest.ts:344]: Those tests are almost an exact copy of [`DynamicFacet`](https://github.com/coveo/search-ui/blob/master/unitTests/ui/DynamicFacet/DynamicFacetTest.ts#L519)'s.

File: src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts (hunk 7 above)
Comment (olamothe) [reply src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts:57]: The classnames are pretty long.

But that's good, that means there will be no nesting needed to customize them :)

File: src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts (hunk 8 above)
Comment (olamothe) [reply src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts:53]: So a screen reader would read the labels in reverse order versus what is actually displayed ? I think that's kind of weird 🤔 ...

File: unitTests/ui/DynamicHierarchicalFacetSearchValueRendererTest.ts [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+          expect(facet.enablePreventAutoSelectionFlag).toHaveBeenCalledTimes(1);
+        });
+
+        it('calls facet.scrollToTop', () => {
+          expect(facet.scrollToTop).toHaveBeenCalledTimes(1);
+        });
+
+        it('calls facet.triggerNewQuery with a function that calls facetValue.logSelectActionToAnalytics', () => {
+          expect(facet.triggerNewQuery).toHaveBeenCalledTimes(1);
+          const [logAnalytics] = <[Function]>(facet.triggerNewQuery as jasmine.Spy).calls.mostRecent().args;
```
Comment (olamothe) [inline unitTests/ui/DynamicHierarchicalFacetSearchValueRendererTest.ts:86]: You can do that kind of args checking with something similar to : 

```
expect(facet.triggerNewQuery).toHaveBeenCalledWith(...params you expect)
//or
expect(facetValue.logSelectActionToAnalytics).toHaveBeenCalledWith(...)
```

File: unitTests/ui/DynamicHierarchicalFacetSearchValueRendererTest.ts [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+      });
+
+      describe('when rendered', () => {
+        let render: HTMLElement;
+
+        beforeEach(() => {
+          render = renderer.render();
+        });
+
+        it('should have an accessible label', () => {
```
Comment (olamothe) [inline unitTests/ui/DynamicHierarchicalFacetSearchValueRendererTest.ts:102]: I think it's better to add an actual accessibility test (that uses axe core) instead of only checking labels and aria properties.

You don't need to remove this since it's already there, but axe-core would help us know if the properties are valid (versus here we just check they exists, without knowing if they "make sense")

File: src/rest/Facet/FacetSearchResponse.ts [hunk 13]
```diff
@@ -5,6 +5,7 @@ export interface IFacetSearchResultValue {
    * **Example:** `ACME Product A`
    */
   displayValue: string;
+  path: string[];
```
Comment (ThibodeauJF) [inline src/rest/Facet/FacetSearchResponse.ts:8]: you may document this

File: gulpTasks/definition.js [hunk 14]
```diff
@@ -54,7 +54,7 @@ function clearEnumVariableDeclaration(match, p1, p2) {
 function getEnumRegex() {
   const enumIdentifier = '(enum [a-zA-Z_$]+\\s{$)';
   const enumDeclaration = '\\n^\\s*[a-zA-Z_$]+ = "[a-zA-Z_$\\s]+",?$';
-  const documentation = '\\n^\\s*[@{}\\[\\]\\w\\/*.,\\s]+$';
+  const documentation = '\\n^\\s*[@{}\\[\\]\\w\\/*.,\\s\\(\\)\\|;]+$';
```
Comment (samisayegh) [inline gulpTasks/definition.js:57]: Why is this change needed?

File: src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearch.ts [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        value: facetValue,
+        element: facetValue.renderedElement
+      };
+    });
+  }
+
+  private onSelectValue({ value }: IComboboxValue) {
+    (<DynamicHierarchicalFacetSearchValueRenderer>value.renderer).selectAction();
+  }
+}
```
Comment (samisayegh) [inline src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearch.ts:62]: Could this file and the next have used the DynamicFacet classes? or were there many differences that creating new files was necessary?

File: src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts (hunk 7 above)
Comment (btaillon) [reply src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts:57]: Yeah, I felt like they were too long but I also felt like shortening them would make them too ambiguous.

File: src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts (hunk 8 above)
Comment (btaillon) [reply src/ui/DynamicHierarchicalFacetSearch/DynamicHierarchicalFacetSearchValueRenderer.ts:53]: Alright, I'll change it to be in the same (large to small) order.

File: unitTests/ui/DynamicHierarchicalFacetSearchValueRendererTest.ts (hunk 11 above)
Comment (btaillon) [reply unitTests/ui/DynamicHierarchicalFacetSearchValueRendererTest.ts:86]: The problem is that I'm testing the function that was passed as a parameter.

File: unitTests/ui/DynamicHierarchicalFacetSearchValueRendererTest.ts (hunk 12 above)
Comment (btaillon) [reply unitTests/ui/DynamicHierarchicalFacetSearchValueRendererTest.ts:102]: Good point, I'll add axe tests.

File: gulpTasks/definition.js (hunk 14 above)
Comment (btaillon) [reply gulpTasks/definition.js:57]: Because `validateTSV1` doesn't support enums with string keys unless the `= 'string value'` part is removed. This was fixed for other enums with this function, but the newly introduced [`FacetSearchType`](https://github.com/coveo/search-ui/blob/master/src/rest/Facet/FacetSearchRequest.ts#L6) has special symbols in its comments that weren't caught in the regex.

File: src/rest/Facet/FacetSearchResponse.ts (hunk 13 above)
Comment (fbeaudoincoveo) [inline src/rest/Facet/FacetSearchResponse.ts:8]: Suggestion:

```
The base path shared by all values returned by a hierarchical facet search request.

**Note:** This property is defined only when a [`basePath`]{@link IFacetSearchRequest.basePath} was specified in the hierarchical facet search request that returned the value.
```

Comment (fbeaudoincoveo) [reply src/rest/Facet/FacetSearchResponse.ts:8]: New suggestion:

```
The hierarchical path to the value.

**Note:** This property is only defined when the facet search request was made against hierarchical values.
```
